rust/projects/parallelism/src/engines/kvs.rs

```diff
@@ -0,0 +1,462 @@
+use std::cell::RefCell;
+use std::collections::BTreeMap;
+use std::fs::{self, File, OpenOptions};
+use std::io::{self, BufReader, BufWriter, Read, Seek, SeekFrom, Write};
+use std::ops::Range;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::{Arc, Mutex};
+
+use crossbeam_skiplist::SkipMap;
+use serde::{Deserialize, Serialize};
+use serde_json::Deserializer;
+
+use super::KvsEngine;
+use crate::{KvsError, Result};
+use std::ffi::OsStr;
+
+const COMPACTION_THRESHOLD: u64 = 1024 * 1024;
+
+/// The `KvStore` stores string key/value pairs.
+///
+/// Key/value pairs are persisted to disk in log files. Log files are named after
+/// monotonically increasing generation numbers with a `log` extension name.
+/// A skip list in memory stores the keys and the value locations for fast query.
+///
+/// ```rust
+/// # use kvs::{KvStore, Result};
+/// # fn try_main() -> Result<()> {
+/// use std::env::current_dir;
+/// use kvs::KvsEngine;
+/// let mut store = KvStore::open(current_dir()?)?;
+/// store.set("key".to_owned(), "value".to_owned())?;
+/// let val = store.get("key".to_owned())?;
+/// assert_eq!(val, Some("value".to_owned()));
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Clone)]
+pub struct KvStore {
+    // directory for the log and other data
+    path: Arc<PathBuf>,
+    // map generation number to the file reader
+    reader: KvStoreReader,
+    index: Arc<SkipMap<String, CommandPos>>,
+    writer: Arc<Mutex<KvStoreWriter>>,
+}
+
+/// A single thread reader.
+///
+/// Each `KvStore` instance has its own `KvStoreReader` and
+/// `KvStoreReader`s open the same files separately. So the user
+/// can read concurrently through multiple `KvStore`s in different
+/// threads.
+struct KvStoreReader {
+    path: Arc<PathBuf>,
+    // generation of the latest compaction file
+    // readers with a generation before safe_point can be closed
```

Comment:
it took me a bit to understand why. Can you add further comment explaining that the compaction generation contains the sum of all operations from previous generations?

Also, I believe previous generations can not only be closed but deleted. If that's correct can you mention that too?

If this comment comes from previous projects then backport any changes.

Reply:
I do see that KvStoreReader doesn't seem responsible for deleting older generations, it's the KvStoreWriter during compaction. This is curious because those deleted generation inodes are going to stick around taking up disk space until every reader is invoked and drops its handles.

I think that's worth noting as a weakness here further down after the "Close readers of stale files" comment.

Reply:
This scheme of holding on to read handles while deleting them probably doesn't work on windows without extra steps. This is probably fine, but we might need to specify somewhere that the example projects are only known to work on Unix.

Reply:
Holding onto the handles though is necessary to make the readers concurrent with the writers I see.

Reply:
On Windows, the deletion will indeed fail immediately after the compaction is done. The stale files are actually deleted in the next compaction. I will add more comments to it.

Reply:
I can't think of a simple way to close those stale file handles right after the compaction, so I leave the work to `KvStoreReader` and let it close them asynchronously. (simple way: not change code a lot and not introduce too many sync mechanisms)

Reply:
I agree this model makes sense here. It's just notable that compaction doesn't immediately free any disk space.